src/lustre/lustreGenRefTypeImpNodes.ml

```diff
@@ -0,0 +1,148 @@
+= fun ctx ty node_id -> match ty with 
+  | RefinementType (pos, (_, id, ty), expr) as ref_type -> 
+    (* Only generate contracts if realizability checking is enabled *)
+    if not (List.mem `CONTRACTCK (Flags.enabled ())) then None else
+    let span = { A.start_pos = Lib.dummy_pos; end_pos = Lib.dummy_pos } in
+    let ty_str = Lib.string_of_t A.pp_print_lustre_type ref_type |> HString.mk_hstring in
+    let gen_node_id = HString.concat2 (HString.mk_hstring (string_of_int !i)) 
+                                      ty_str in
+    i := !i + 1;
+    let is_extern = true in
+    let params = [] in 
+    let vars = AH.vars_without_node_call_ids expr in
+    let inputs = SI.diff vars (SI.singleton id) |> SI.elements in
+    let inputs = List.filter_map (fun id -> 
+      let ty = Ctx.lookup_ty ctx id |> unwrap in
+      let ty = Chk.expand_type_syn_reftype_history ctx ty |> unwrap_res in
+      let is_const = Ctx.member_val ctx id in
+      let call_params = (match node_id with 
+      | Some node_id -> Ctx.lookup_node_param_ids ctx node_id |> unwrap 
+      | None -> []
+      ) in
+      let is_global_const = is_const && not (List.mem id call_params) in
+      if is_global_const 
+      then None 
+      else Some (pos, id, ty, A.ClockTrue, is_const)
+    ) inputs in
+    let outputs = [(pos, id, ty, A.ClockTrue)] in
+    let locals = [] in 
+    let node_items = [A.AnnotMain(pos, true)] in 
+    (* Add assumption for each variable with a refinement type in 'expr' *)
+    let assumptions = List.filter_map (fun (_, id, _, _, _) -> 
+      let ty = Ctx.lookup_ty ctx id |> unwrap in 
+      match ty with 
+        | A.RefinementType (_, (_, id2, _), expr) -> 
+          let expr =  (AH.substitute_naive id2 (Ident(pos, id)) expr) in
+          Some (A.Assume (pos, None, false, expr))
+        | _ -> None 
+    ) inputs in 
+    (* Add guarantee for 'expr' *) 
+    let guarantee = A.Guarantee (pos, None, false, expr) in
+    let contract = Some (guarantee :: assumptions) in
+    Some (NodeDecl (span, (gen_node_id, is_extern, params, inputs, outputs, locals, node_items, contract)))
+  | _ -> None
+
+let gen_imp_nodes: Ctx.tc_context -> A.declaration list -> A.declaration list 
+= fun ctx decls -> 
+  List.fold_left (fun acc decl -> 
+    match decl with 
+    | A.TypeDecl (_, AliasType (_, _, ty))
+    | A.ConstDecl (_, FreeConst (_, _, ty))
+    | A.ConstDecl (_, TypedConst (_, _, _, ty)) -> 
+      (match (ref_type_to_contract ctx ty None) with  
+      | None -> decl :: acc
+      | Some decl2 -> decl :: decl2 :: acc)
+    | A.TypeDecl (_, FreeType _)
+    | A.ConstDecl (_, UntypedConst _) -> decl :: acc
+    | A.NodeDecl (span, ((id, extern, params, inputs, outputs, locals, node_items, contract) as decl2)) -> 
+      let decl = 
+        if not (List.mem `CONTRACTCK (Flags.enabled ())) then decl else
+        (* To prevent slicing, we mark generated imported nodes as main nodes *)
```

Comment:
In fact, here you are marking the original nodes as main nodes. Is it really required? If so, update the comment to explain it, and move this comment to `node_decl_to_contracts` and `ref_type_to_contract`.

Reply:
Addressed. It is not necessary in this instance, but the other main annotations are necessary. I removed the main annotation and comment in this location, and included the comment in the spots where the main annotations are actually needed.